server/core/src/rodsAgent.cpp

```diff
@@ -192,437 +224,529 @@ void cleanup()
+        close(listen_tmp_socket);
+
+        if (agent_pid == 0) {
+            // This is the child process.
+            // Agent logic starts outside of the while-loop.
+            break;
+        }
+        else if (agent_pid > 0) {
+            // This is the parent process.
+            // Clean up and prepare to fork more agents upon request.
+            close(conn_tmp_socket);
+        }
+        else if (agent_pid < 0) {
+            log_agent_factory::critical("Agent factory failed to fork agent. Shutting down agent factory ...");
+
+            close(client_socket);
+            close(listen_socket);
+
+            return SYS_FORK_ERROR;
+        }
+    } // Agent factory main loop.
+
+    //
+    // This is where the agent logic actually begins.
+    //
+
+    int status{};
+
+    try {
+        log::set_server_type("agent");
+
+        // Reload irods_environment.json and server_config.json for the newly forked agent process.
+        irods::environment_properties::instance().capture();
+        irods::server_properties::instance().capture();
+
+        log_agent::trace("Agent forked. Initializing ...");
+
+        close(listen_socket);
+
+        // Restore signal dispositions for agents.
+        std::signal(SIGABRT, SIG_DFL);
+        std::signal(SIGINT,  SIG_DFL);
+        std::signal(SIGHUP,  SIG_DFL);
+        std::signal(SIGTERM, SIG_DFL);
+        std::signal(SIGCHLD, SIG_DFL);
+        std::signal(SIGUSR1, SIG_DFL);
+        std::signal(SIGPIPE, SIG_DFL);
+
+        status = receiveDataFromServer(conn_tmp_socket);
+        if (status < 0) {
+            log_agent::error("receiveDataFromServer failed [error_code=[{}]].", status);
+        }
+
+        close(conn_tmp_socket);
+
+        set_eviction_age_for_dns_and_hostname_caches();
+        set_log_levels_for_all_log_categories();
+
+        if (const auto err = setRECacheSaltFromEnv(); !err.ok()) {
+            log_agent::error("rodsAgent::main: Failed to set RE cache mutex name\n%s", err.result());
+            return SYS_INTERNAL_ERR;
         }
     }
+    catch (const irods::exception& e) {
+        log_agent::error("Agent initialization error: {}", e.what());
+        return e.code() == -1 ? SYS_UNKNOWN_ERROR : e.code();
+    }
+    catch (const std::exception& e) {
+        log_agent::error("Agent initialization error: {}", e.what());
+        return SYS_LIBRARY_ERROR;
+    }
+
+    RsComm rsComm{};
+
+    log::set_error_object(&rsComm.rError);
+    irods::at_scope_exit release_error_stack{[] { log::set_error_object(nullptr); }};
 
-    memset( &rsComm, 0, sizeof( rsComm ) );
-    rsComm.thread_ctx = ( thread_context* )malloc( sizeof( thread_context ) );
+    //std::memset(&rsComm, 0, sizeof(RsComm));
+    rsComm.thread_ctx = static_cast<thread_context*>(std::malloc(sizeof(thread_context)));
 
-    status = initRsCommWithStartupPack( &rsComm, nullptr );
+    status = initRsCommWithStartupPack(&rsComm, nullptr);
```

Comment:
Apparently, this function treats a socket fd of 0 as invalid which is incorrect. 0 is in fact a valid socket fd.

The socket fd is retrieved via an environment variable and parsed using `std::atoi()`. `std::atoi()` can return 0 if it cannot parse the input.

Because of this, the server never stores the IP address in the `RsComm` causing  `iticket` tests to fail.

Reply:
The values stored in the environment need to be logged as debug.

We should consider logging inputs and outputs of functions for debug purposes. Perhaps as compile-time only log messages.

Reply:
There's nothing actionable here.

Marking as resolved.